zcash_client_sqlite/src/wallet/common.rs

```diff
@@ -116,6 +116,163 @@ where
 
 #[allow(clippy::too_many_arguments)]
 pub(crate) fn select_spendable_notes<P: consensus::Parameters, F, Note>(
+    conn: &Connection,
+    params: &P,
+    account: AccountUuid,
+    target_value: TargetValue,
+    anchor_height: BlockHeight,
+    exclude: &[ReceivedNoteId],
+    protocol: ShieldedProtocol,
+    to_spendable_note: F,
+) -> Result<Vec<ReceivedNote<ReceivedNoteId, Note>>, SqliteClientError>
+where
+    F: Fn(&P, &Row) -> Result<Option<ReceivedNote<ReceivedNoteId, Note>>, SqliteClientError>,
+{
+    match target_value {
+        TargetValue::MaxSpendable => select_maximum_spendable_notes(
+            conn,
+            params,
+            account,
+            anchor_height,
+            exclude,
+            protocol,
+            to_spendable_note,
+        ),
+        TargetValue::MinValue(zats) => select_minimum_spendable_notes(
+            conn,
+            params,
+            account,
+            zats,
+            anchor_height,
+            exclude,
+            protocol,
+            to_spendable_note,
+        ),
+    }
+}
+
+/// selects all the spendable notes available above ``MARGINAL_FEE`` amount from a given account
+/// and specified shielded protocols excluding the ones in the exclude slice.
+///
+/// - Implementation details:
+///   - Notes with individual value *below* the ``MARGINAL_FEE`` will be ignored
+///   - Note spendability is determined using the `anchor_height`
+#[allow(clippy::too_many_arguments)]
+fn select_maximum_spendable_notes<P: consensus::Parameters, F, Note>(
+    conn: &Connection,
+    params: &P,
+    account: AccountUuid,
+    anchor_height: BlockHeight,
+    exclude: &[ReceivedNoteId],
+    protocol: ShieldedProtocol,
+    to_spendable_note: F,
+) -> Result<Vec<ReceivedNote<ReceivedNoteId, Note>>, SqliteClientError>
+where
+    F: Fn(&P, &Row) -> Result<Option<ReceivedNote<ReceivedNoteId, Note>>, SqliteClientError>,
+{
+    let birthday_height = match wallet_birthday(conn)? {
+        Some(birthday) => birthday,
+        None => {
+            // the wallet birthday can only be unknown if there are no accounts in the wallet; in
+            // such a case, the wallet has no notes to spend.
+            return Ok(vec![]);
+        }
+    };
+
+    let (table_prefix, index_col, note_reconstruction_cols) = per_protocol_names(protocol);
+    if unscanned_tip_exists(conn, anchor_height, table_prefix)? {
+        return Ok(vec![]);
+    }
```

Comment:
This seems like the wrong place to be performing this check, but I think it's copy-pasta from one of the other existing implementations. It is definitely in the wrong place for the case where we change how spendability is checked to allow different accounts in the wallet to have different spendability based upon the state of the note commitment tree; please open an issue so that this doesn't get missed when we go to make these changes.

Reply:
https://github.com/zcash/librustzcash/issues/1796 and added it to the code